### ISSUE TITLE
KafkaConsumer: Potential to lose connection?

### ISSUE THREAD

**Issue by ddebrunner** [IBMStreams/streamsx.kafka#60]
Under detecting **Detecting Consumer Failures** in `KafkaConsumer` javadoc it says:

_Underneath the covers, the poll API sends periodic heartbeats to the server; when you stop calling poll (perhaps because an exception was thrown), then no heartbeats will be sent. If a period of the configured session timeout elapses before the server has received a heartbeat, then the consumer will be kicked out of the group and its partitions will be reassigned._

I don't see anything in the operator that handles this case, so is it possible that while the operator is checkpointing or resetting (or otherwise not polling) that the consumer is kicked out, and thus no more messages are consumed?

It seems that the operator should register a `ConsumerRebalanceListener` callback when subscribing and then react to `onPartitionsRevoked` by reseting the consumer (or creating a new one)?

https://kafka.apache.org/0100/javadoc/index.html?org/apache/kafka/clients/consumer/KafkaConsumer.html

**Comment by RolefH** [comment]
The java doc of KafkaConsumer seems not be up-to-date. Since Kafka client 0.10.1.0, the heartbeat is separated from poll(). [KIP62](https://cwiki.apache.org/confluence/display/KAFKA/KIP-62%3A+Allow+consumer+to+send+heartbeats+from+a+background+thread#KIP-62:Allowconsumertosendheartbeatsfromabackgroundthread-TestPlan) and the consumer property max.poll.interval.ms controls the maximum time between consecutive calls of poll() before the consumer is treated being dead. The default value for this property is 5 minutes, while the default session timeout is 10 seconds.
Implementing a `ConsumerRebalanceListener` will not help since the listener is only notified if a consumer *subscribes*, but not when a consumer *assigns* to a topic. It is also unclear, whether the Callback is also called by the group coordinator at those consumers, that are kicked out.

In a first step I propose to
1. measure the time between calls to poll()
1. If the time between two poll()'s is higher than `max.poll.interval.ms` give a warning in log and/or trace
1. Do nothing, because the next call to poll() will make the make the consumer re-join the consumer group.

**Comment by ddebrunner** [comment]
There must be some callback when the consumer is dead, otherwise how does a client know?

**Comment by RolefH** [comment]
I tested the behavior of the kafka consumer in case that it stopped polling for more than `max.poll.interval.ms`. The consumer thread stops polling when

a) control port tuple is processed (should really not take longer than `max.poll.interval.ms`)
b) on checkpoint
c) when the internal message queue is full, for example when the downstream application is congested (this was also my test scenario. I have tested with max.poll.interval.ms=60000, internal queue size of 5000 and blocking times between 5 and 10 minutes.)
d) when the KafkaSource triggers the CR

If the group coordinator is involved, i.e. when the operator _subscribes_ to topics, another consumer in the same group takes over, and the blocked consumer joins the group again when it continues polling. If there is only one consumer in the group, for example with default random groupId, the  consumer continues receiving messages when it continues polling.
When a consumer _assigns_ to topic partitions, no group coordinator is involved, i.e. the load is not re-balanced when a consumer stops polling or when it stops sending heartbeats. In this case the consumer continues to receive messages from its partition(s), when it continues polling.
It is also possible to stop and restart the kafka broker in between. Producer and consumers reconnected.

One thing must be done to make these scenarios work: Auto-commit at consumer side must be disabled. The messages that have been put into the internal message queue must be committed by the consumer thread immediately (remember, that the consumer API is no thread-safe). Otherwise we would see duplicated messages in the load balancing scenario.
